packages/xpetra/sup/CMakeLists.txt

```diff
@@ -12,12 +12,63 @@ TRIBITS_CONFIGURE_FILE(${PACKAGE_NAME}_config.hpp)
 # SRC
 #
 
+##
+## XPETRA_ETI_CPP_SOURCES collects all filenames of the Xpetra ETI *.cpp files
+## These are later added to the global list of source files (SOURCES)
+##
+SET(SOURCES "")
+SET(XPETRA_ETI_CPP_SOURCES "")
+
+##
+## only build ETI *.cpp files if ETI is enabled.
+##
+IF(${PACKAGE_NAME}_ENABLE_EXPLICIT_INSTANTIATION)
+  ##
+  ## XPETRA_SC_LO_GO_NO_ETI_CLASSES is a list of all class names which ETI 
+  ## *.cpp files have to be generated for.
+  ##
+  GLOBAL_SET(XPETRA_SC_LO_GO_NO_ETI_CLASSES  )
+  # include the list of all Xpetra classes that should be ETI'd
+  INCLUDE(Utils/ExplicitInstantiation/ETI_SC_LO_GO_NO_classes.cmake)
+
+  # include the function XPETRA_PROCESS_ETI_TEMPLATE
+  include(${CMAKE_CURRENT_SOURCE_DIR}/../src/Utils/ExplicitInstantiation/xpetra-process-eti-template.cmake)
+
+  #FIXME
+  #ninja: warning: multiple rules generate
+  #/scratch/jhu/trilinos/Trilinos/packages/xpetra/src/Utils/ExplicitInstantiation/Xpetra_ETI_SC_LO_GO_NO.tmpl. builds
+  #involving this target will not be correct; continuing anyway [-w dupbuild=warn]
+
+
+  # generate the cpp files and record the file names
+  XPETRA_PROCESS_ETI_TEMPLATE(
+    "${XPETRA_SC_LO_GO_NO_ETI_CLASSES}"
+    ${CMAKE_CURRENT_SOURCE_DIR}/../src/Utils/ExplicitInstantiation/Xpetra_ETI_SC_LO_GO_NO.tmpl
+    "Utils/ExplicitInstantiation/CLASS_FILE_NAME.cpp"
+    Xpetra_ETI_SC_LO_GO_NO_SRC)
+  LIST(APPEND XPETRA_ETI_CPP_SOURCES ${Xpetra_ETI_SC_LO_GO_NO_SRC})
+
+  # At this point, XPETRA_ETI_CPP_SOURCES contains the Xpetra classes
+  # that will be ETI'd.
+
+  APPEND_GLOB(SOURCES ${XPETRA_ETI_CPP_SOURCES})
+
+  SET(SOURCES ${XPETRA_ETI_CPP_SOURCES})
+  # grab any hand rolled ETI source files
+  APPEND_GLOB(SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/Utils/ExplicitInstantiation/*.cpp)
+  
+ENDIF() # if XPETRA_ENABLE_EXPLICIT_INSTANTIATION
+
 SET(HEADERS "")
 APPEND_GLOB(HEADERS */*.hpp)
-APPEND_GLOB(HEADERS */*.cpp)
+APPEND_GLOB(HEADERS */*.cpp) #JHU 25-Apr-2019 FIXME is this correct?
```

Comment:
Wouldn't you want `.cpp` files to go into `SOURCES`, so they can go into the list of files to compile into the library (see `TRIBITS_ADD_LIBRARY` invocation below)?